### ISSUE TITLE
[clang-tidy] Types of fixed size and access cause warning cppcoreguidelines-pro-bounds-avoid-unchecked-container-access

### ISSUE THREAD

**Issue by DaAwesomeP** [llvm/llvm-project#198176]
It seems that `cppcoreguidelines-pro-bounds-avoid-unchecked-container-access` triggers when accessing fixed-bounds types with a fixed index. The following example produces a warning for `my_var[0]` and all 3 instances of `my_var[1]`:

```cpp
std::array<std::int32_t, 2> my_var = {1, 1};
if (some_var > 0) {
    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
    file_stream.read(reinterpret_cast<char *>(my_var.data()), sizeof(my_var[0]));
}
if (some_var > 1)
    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
    file_stream.read(reinterpret_cast<char *>(&my_var[1]), sizeof(my_var[1]));
    another_var *= my_var[1];
}
```
These should not produce any warnings since `my_var` is of fixed size (2) and it is being accessed with a fixed integer (not a variale index) that is within bounds. I left the unrelated code in this example in case that is a part of the bug.

The [C++ Core Guidelines SL.con.3](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#rsl-bounds) provide a similar example as acceptable for array access (here my `my_var` is like the `a` example):
```cpp
void f(std::vector<int>& v, std::array<int, 12> a, int i)
{
    v[0] = a[0];        // BAD
    v.at(0) = a[0];     // OK (alternative 1)
    at(v, 0) = a[0];    // OK (alternative 2)

    v.at(0) = a[i];     // BAD
    v.at(0) = a.at(i);  // OK (alternative 1)
    v.at(0) = at(a, i); // OK (alternative 2)
}
```
Clang-tidy version: `lvm-toolchain-noble-22/main amd64 clang-tidy-22 amd64 1:22.1.6~++20260514073821+fc4aad7b5db3-1~exp1~20260514073942.73`

**Comment by chrchr-github** [comment]
https://godbolt.org/z/1Pbh1dd1r

Edit: `std::inplace_vector` has a known capacity, but variable size.

**Comment by DaAwesomeP** [comment]
Hi @chrchr-github thank you for confirming the issue. May this issue be tagged as confirmed?

**Comment by chrchr-github** [comment]
> Hi [@chrchr-github](https://github.com/chrchr-github) thank you for confirming the issue. May this issue be tagged as confirmed?

Maybe @firewave or @zeyi2 can help with that.
